backends/p4tools/common/CMakeLists.txt

```diff
@@ -41,6 +41,8 @@ add_p4tools_library(p4tools-common ${P4C_TOOLS_COMMON_SOURCES})
 target_link_libraries(
   p4tools-common
   PUBLIC ${P4TOOLS_Z3_LIB}
+  PRIVATE frontend
+  PRIVATE ir-generated
```

Comment:
Hmm I am only realizing this now, but `ir-generated` should never show up outside the `ir` folder. It should be encapsulated by the `ir` library, which effectively exports all ir-related utilities. If this is necessary we might be doing something wrong. This might also explain why we sometimes get the `ir/ir-generated.h` includes. This should not happen.

Reply:
Let me check what is going on. Here I was essentially transitioning from existing `add_dependency` to proper library

Reply:
This looks like a common pattern that was Ctrl+C / Ctrl-V all around the backends:
```
ubpf/CMakeLists.txt
64:add_dependencies(p4c-ubpf ir-generated frontend)

p4tools/modules/testgen/test/z3-solver/constraints.cpp
8:#include "ir/ir-generated.h"

p4tools/modules/testgen/targets/bmv2/expr_stepper.cpp
21:#include "ir/ir-generated.h"

p4test/CMakeLists.txt
31:add_dependencies(p4test ir-generated frontend)

graphs/CMakeLists.txt
33:target_link_libraries(p4cgraphs ir-generated frontend)
36:target_link_libraries (p4c-graphs p4cgraphs ir-generated frontend ${P4C_LIBRARIES} ${P4C_LIB_DEPS})

ebpf/CMakeLists.txt
126:add_dependencies(p4c-ebpf ir-generated frontend)

bmv2/CMakeLists.txt
86:target_link_libraries(bmv2backend ir-generated frontend)
```

Reply:
As far as I can see, `ir-generated` is not a part of `ir`. It is setup in top-level CMakeLists:
```
# Component libraries: must be defined before being used in the back end executables.
set (P4C_LIBRARIES controlplane midend frontend ir p4ctoolkit ir-generated)
```
and more rules further on.

Probably something should be refactored. I tend to agree that these `-gen` targets should be internal to corresponding libraries.

Reply:
No you are right, I misremembered. There are two parts here. `ir-generated.h` is always meant to be encapsulated by `ir.h`. It used to be the case that one caused compilation errors when including the `ir-generated` header, but I cleaned that up at some point. Now it doesn't cause issues anymore but one should still include `ir.h`. I will clean up the `ir-generated.h` includes, some of them are caused by me and clang-tidy's include fixer :disappointed:. I need to configure that one right. 

The CMake target itself is actually separate and each back end links to it separately. Not a big fan of it either, but that is partially necessary because each back end can add their own IR sources to the target, like the DPDK target does: 
https://github.com/p4lang/p4c/blob/main/backends/dpdk/CMakeLists.txt#L99

We can keep thing as is for now.